nexus/reconfigurator/blippy/src/checks.rs

```diff
@@ -0,0 +1,1561 @@
+        datasets_config.datasets.retain(|_, dataset| {
+            let matches_durable = (dataset.pool
+                == *durable_zone.zone_type.durable_zpool().unwrap())
+                && (dataset.kind
+                    == durable_zone.zone_type.durable_dataset().unwrap().kind);
+            let root_dataset = root_zone.filesystem_dataset().unwrap();
+            let matches_root = (dataset.pool == *root_dataset.pool())
+                && (dataset.kind == *root_dataset.dataset());
+            !matches_durable && !matches_root
+        });
+
+        let expected_notes = [
+            Note {
+                severity: Severity::Fatal,
+                kind: Kind::Sled {
+                    sled_id: *sled_id,
+                    kind: SledKind::ZoneMissingFilesystemDataset {
+                        zone: root_zone.clone(),
+                    },
+                },
+            },
+            Note {
+                severity: Severity::Fatal,
+                kind: Kind::Sled {
+                    sled_id: *sled_id,
+                    kind: SledKind::ZoneMissingDurableDataset {
+                        zone: durable_zone,
+                    },
+                },
+            },
+        ];
+
+        let report =
+            Blippy::new(&blueprint).into_report(BlippyReportSortKey::Kind);
+        eprintln!("{}", report.display());
+        for note in expected_notes {
+            assert!(
+                report.notes().contains(&note),
+                "did not find expected note {note:?}"
+            );
+        }
+
+        logctx.cleanup_successful();
+    }
+
+    #[test]
+    fn test_sled_missing_datasets() {
+        static TEST_NAME: &str = "test_sled_missing_datasets";
+        let logctx = test_setup_log(TEST_NAME);
+        let (_, _, mut blueprint) = example(&logctx.log, TEST_NAME);
+
+        // Pick one sled and remove its blueprint_datasets entry entirely.
+        let removed_sled_id = *blueprint
+            .blueprint_datasets
+            .keys()
+            .next()
+            .expect("at least one sled");
+        blueprint
+            .blueprint_datasets
+            .retain(|&sled_id, _| sled_id != removed_sled_id);
+
+        let report =
+            Blippy::new(&blueprint).into_report(BlippyReportSortKey::Kind);
+        eprintln!("{}", report.display());
+        let mut found_sled_missing_note = false;
+        for note in report.notes() {
+            if note.severity == Severity::Fatal {
+                match &note.kind {
+                    Kind::Sled {
+                        sled_id,
+                        kind: SledKind::SledMissingDatasets { .. },
+                    } if *sled_id == removed_sled_id => {
+                        found_sled_missing_note = true;
+                    }
+                    _ => (),
+                }
+            }
+        }
+        assert!(found_sled_missing_note, "found sled missing datasets note");
```

Comment:
Nitpick; this isn't an expect -- we'll see this message if the note **isn't** found, right?

Reply:
To be super clear, I think the error message should be `Did not find sled with missing datasets note`